docs/WinRTAPIContracts.md

```diff
@@ -0,0 +1,76 @@
+# Windows App SDK API Contracts
+
+## Summary
+Customers trust Microsoft that the products we make are sound, and will work.  Windows App SDK (WASDK) has recently released version 1.0.  As WASDK evolvs and more developers use WASDK in their own projects WASDK needs a way to guarantee that the API’s that ship are stable and that API versions are well documented.
+To keep customers’ trust in WASDK, and to make sure the WASDK API’s are
+1.  Binary compatible
+2.  Stable
+3.  Customizable so developers can include only the parts they want
+
+Contracts, and contract versions should be added to WASDK.
+
+## Architecture Decisions
+1. Contract boundaries
+Contracts will be split among the winmd boundaries.  This is a good medium between having 1 contract for everything, and 1 contract per class.
+2. Contract numbering.
+Contracts will be numbered with MAJOR.MINOR.  Any update to the API is a MAJOR change.  Any servicing update is a MINOR change.
```

Comment:
Contract versions already have semantics, but since they were originally designed for platform use, they have no notion of a "breaking" change semantic. For now, I do like the idea of using the minor version number to indicate contract increments within the 1.* Windows App SDK series, but I think at this time we should leave it as an open issue as to how we might handle breaking changes. The contract numbering is primarily to indicate ordering of when things are introduced, but the numbers themselves are mostly meaningless outside of the inner workings of the type system. There is an "is contract present" API but the type / API checks are preferred over contract checks (IsContractPresent further suggests that within a contract there's never breaking changes, breaking change might imply new contract name).

Reply:
I think that each release of the Windows App SDK should increment the *major* version, but not because we're using API contracts to imply a degree of change or compat (aka semantic versioning).

API contracts should increment the major version because, like @BenJKuhn indicated, the contract versioning indicates a strict ordering of API changes, but the API contract format in WinRT winmds has only major+minor, not major+minor+patch+build - we've already released multiple 0.8.* versions of Windows App SDK, and shouldn't be surprised to see a 1.0.1 release. For this reason alone, it's a **non-goal** for API contract versions to attempt to match Windows App SDK versions.

Incrementing just the major version removes any ambiguity about "which version is next". When deciding what to number the successor to contract 1.0, is it 1.1 or 2.0? Having a clear simple rule is a virtue here.

Another reason is that someday we may need to modify a contract version in a servicing branch, and we need to be able to do it in a way that maintains the strict ordering in contract versions. For example, say we've released Windows App SDK 1.1 and 1.2 containing `FooContract` 1.1 and 1.2, respectively. If we needed to make a change to `FooContract` 1.1, we wouldn't be able to version it properly because we already used up all of our precision/granularity when we released `FooContract` 1.2. We avoid that dilemma by incrementing only major versions of API contracts in stable releases, and keeping the minor version in our pocket as an emergency pressure relief.